### PR TITLE
Implement phased harness hardening (issues #11-#16)

### LINKED ISSUE #11: [Phase 0] Baseline and engineering guardrails
## Summary
Create immediate safety nets so future harness work is validated automatically and consistently.

## Why
The project currently lacks CI workflow enforcement. This makes regressions likely and slows iteration quality.

## Scope
- Add GitHub Actions workflow for:
  - `gofmt -l`
  - `go vet ./...`
  - `go test ./...`
- Add a local helper (`Makefile` targets or script) that mirrors CI checks.
- Document contribution workflow in README.
- Capture a baseline quality snapshot:
  - Test pass/fail
  - Package-level test durations
  - Any existing latency metrics availability

## Out of scope
- Full eval harness
- Benchmark dashboards
- Security sandbox implementation

## Deliverables
- `.github/workflows/ci.yml`
- Local check command (`make ci` or equivalent)
- README update for contributor checks
- Baseline metrics artifact or markdown report

## Acceptance criteria
- CI runs on push and PR.
- CI is required for merge (branch protection setup documented if not automated).
- Running the local command produces equivalent results to CI.

## Dependencies
- None

## Risks
- CI runtime may be noisy initially due to flaky tests.
- Mitigation: mark flaky tests and open follow-up issues immediately.

## Suggested labels
`phase:0`, `type:infrastructure`, `area:ci`, `priority:P0`

## Milestone
`Agent Harness Hardening`

### LINKED ISSUE #12: Phase 1: Execution Safety Hardening
## Summary
Harden tool execution boundaries so prompt injection or model mistakes cannot cause broad filesystem or shell side effects.

## Problem
Current safety controls are primarily policy/denylist based. The harness needs stronger enforcement boundaries for file access and command execution.

## Scope
- Add workspace path jail for file tools.
- Replace broad shell execution with allowlist-based command policy.
- Expand approval checks and bypass-resistance tests.

## Deliverables
- Path normalization + workspace-root guard for read/write tools.
- Command execution adapter that validates binary + args before execution.
- Security regression tests covering blocked patterns and obfuscation attempts.
- Updated docs describing allowed command surface and approval flow.

## Out of scope
- Full container or VM sandbox runtime (tracked in later phase if needed).

## Tasks
- [ ] Define `allowedRoots` policy (workspace default, optional explicit extensions).
- [ ] Refactor file tool path resolution to reject out-of-root traversal.
- [ ] Introduce command parser/validator and allowlist (e.g., `ls`, `pwd`, `go test`, etc. configurable).
- [ ] Ensure command approvals are tied to normalized command identity.
- [ ] Add tests for traversal, shell metacharacter tricks, chained commands, and approval replay edge cases.
- [ ] Document policy configuration in README.

## Acceptance criteria
- Any path outside allowed roots is rejected with a clear error.
- Any non-allowlisted command is denied or requires explicit approval based on policy.
- Security test suite passes in CI.

## Dependencies
- Phase 0 CI gate must exist first.

## Risks
- Over-restrictive policies may block legitimate workflows.
- Mitigation: configurable allowlist + explicit approval override with audit trail.

## Suggested labels
`harness`, `security`, `tools`, `phase-1`

### LINKED ISSUE #13: Phase 2: Build evaluation harness v1 (deterministic scenarios + CI scoring)
## Summary
Introduce a first-class evaluation harness so quality is measurable and regressions are detectable. This phase defines the eval case format, adds an initial corpus of scenarios, and runs eval scoring in CI on every PR.

## Problem
Current development lacks a repeatable quality measurement system. Improvements are difficult to compare objectively, and regressions can ship unnoticed.

## Scope
- In scope:
  - Eval case schema (JSON or YAML) capturing:
    - input prompt
    - optional setup context
    - expected behavioral assertions (tool usage, safety constraints, output properties)
  - Eval runner command (e.g. `go run ./cmd/eval` or `go test` package-based harness)
  - Initial corpus (at least 20 scenarios) across:
    - tool-calling correctness
    - command safety boundaries
    - memory retrieval and fallback behavior
    - compaction-sensitive conversation flow
  - Machine-readable output (JSON summary + per-case result)
  - CI integration to run eval suite and surface failures in PR checks
- Out of scope:
  - Full benchmark dashboards (Phase 5)
  - Large-scale production traffic replay

## Deliverables
- Eval schema docs and examples.
- Eval runner with deterministic execution hooks where possible.
- Initial eval dataset committed to repo.
- CI workflow/job for eval scoring.
- Developer docs for running eval locally.

## Acceptance criteria
- Running a single command locally executes all evals and produces a summary report.
- CI fails when any required eval assertion fails.
- Results include enough detail to understand which behavior regressed.
- Eval corpus covers key safety and tool-routing behaviors.

## Dependencies
- Requires Phase 0 CI baseline.
- Coordinates with Phase 1 for safety-specific assertions.

## Suggested labels
- `type:enhancement`
- `area:evals`
- `priority:P0`

## Suggested milestone
Agent Harness v1

## Implementation checklist
- [ ] Define eval case schema and fixtures directory structure.
- [ ] Implement eval runner and assertion engine.
- [ ] Add at least 20 initial scenarios with expected assertions.
- [ ] Add CI job and artifact upload for eval results.
- [ ] Document workflow in README or docs.

## Risks / notes
- Model nondeterminism can cause flaky assertions. Favor behavioral/property assertions over exact string matching.
- Keep seed and metadata where feasible to aid reproducibility.

### LINKED ISSUE #14: Phase 3: Observability and run artifacts
## Summary
Instrument the harness so behavior is inspectable at turn and tool granularity, and preserve machine-readable artifacts for debugging and analysis.

## Problem
Current logging is mostly unstructured. Regressions and production failures are difficult to investigate without rerunning scenarios manually.

## Scope
- Introduce structured event schema for:
  - agent turn start/end
  - tool call start/end
  - tool error categories
  - compaction/memory retrieval decisions
- Add metrics counters/histograms for:
  - turn latency
  - tool latency by tool name
  - tool failure rate
  - fallback parser usage
- Persist run artifacts:
  - normalized transcript
  - tool execution summary
  - run metadata (session, gateway, model, timestamps)
- Add docs for artifact format and retention policy.

## Deliverables
- `telemetry/` package (or equivalent) with event + metric helpers.
- Structured logs emitted from core loop and tool executor.
- Artifact writer module with configurable output path.
- Developer docs for local troubleshooting using artifacts.

## Acceptance criteria
- A failed run can be diagnosed from artifacts/logs without reproducing interactively.
- Telemetry includes stable machine-readable keys suitable for dashboards.
- Feature flag/config controls verbosity and artifact retention.

## Dependencies
- Depends on Phase 2 schema decisions for eval/run metadata compatibility.

## Risks
- Logging volume and overhead increase.
- Inconsistent field names without schema governance.

## Suggested labels
- `type:enhancement`
- `area:observability`
- `area:telemetry`
- `priority:p1`

## Milestone
Agent Harness v1

### LINKED ISSUE #15: [Phase 4] Reliability and regression depth
## Summary
Expand test depth with integration and failure-injection coverage for the most critical agent workflows.

## Why this matters
- Unit tests are not enough to validate emergent behavior in multi-round tool loops.
- Queueing, memory fallback, and gateway interactions can fail in ways that only appear in end-to-end execution.

## Scope
- Add integration tests for representative user flows:
  - Multi-round tool use with final response
  - Approval-required command denial/approval/execute cycle
  - Wallet approval path (if wallet enabled)
- Add failure-injection tests:
  - Embedding service unavailable
  - Compaction returns empty/invalid payload
  - Queue worker panic and recovery behavior
- Add deterministic fixtures for session/memory/reminder stores.

## Out of scope
- Distributed load testing infrastructure.
- Large-scale benchmark trend analysis (Phase 5).

## Deliverables
- Integration test suite executable via `go test` and CI.
- Failure-mode regression tests for known fragile paths.
- Test documentation for local and CI execution.

## Acceptance criteria
- Critical workflows have integration coverage with deterministic assertions.
- Known failure paths produce expected fallback behavior and logs.
- Test flake rate is low and documented if non-zero.

## Dependencies
- Requires Phase 1 safety controls and Phase 3 instrumentation to maximize diagnostic value.

## Suggested labels
- `type:testing`
- `area:agent-loop`
- `priority:P1`

## Suggested milestone
- `Agent Harness v1`

## Task checklist
- [ ] Create integration test harness utilities and shared fixtures.
- [ ] Implement high-value scenario tests for tool loop and approvals.
- [ ] Implement failure-injection tests for memory/compaction/queue.
- [ ] Add CI job partitioning if runtime is too long.
- [ ] Document expected runtime and stability targets.

### LINKED ISSUE #16: [Phase 5] Benchmarking and release governance
## Summary
Institutionalize the agent improvement loop with trend reporting and release gates based on quality, safety, and latency thresholds.

## Why this matters
- Improvement requires stable measurement and release discipline.
- Teams need explicit ship criteria tied to benchmark outcomes.

### In scope
- Benchmark suite runner and reporting format.
- Release checklist with gating thresholds.
- Changelog template for quality-impact reporting.

### Out of scope
- Full product analytics platform.
- Replacing existing release tooling beyond harness gates.

## Deliverables
- `benchmarks/` runner and configuration.
- Trend report artifact (JSON + markdown summary).
- Release gate checklist and documentation.

## Implementation tasks
- [ ] Define benchmark dimensions:
  - Task success rate
  - Tool correctness rate
  - Safety violation rate
  - p50/p95 latency
- [ ] Build benchmark runner over eval corpus + integration tasks.
- [ ] Generate trend artifacts per run (machine + human readable).
- [ ] Add release gate script checking thresholds.
- [ ] Add release checklist to docs/README with fail/waive policy.
- [ ] Add changelog section template: "Agent Quality Impact".

## Acceptance criteria
- [ ] Benchmarks run on demand and in CI release pipeline.
- [ ] Threshold failures clearly block release unless explicitly waived.
- [ ] Trend reports are retained and comparable between runs.

## Dependencies
- Phase 2 eval outputs.
- Phase 3 telemetry outputs.
- Phase 4 reliability tests.

## Risks
- Thresholds may be unrealistic initially.
  - Mitigation: start with baseline-derived thresholds and tighten iteratively.

## Suggested labels
- `type:enhancement`
- `area:benchmarking`
- `area:release`
- `priority:p2`

## Suggested milestone
- `Harness Hardening v1`

### PR DESCRIPTION
**Pull request by fabriziogianni7**
## Summary
This PR implements the full sequential harness-hardening roadmap across issues #11 through #16, with each phase completed and validated before moving to the next.

### Completed phases
- **#11 Phase 0 – Baseline and guardrails**
  - Added CI workflow (`gofmt`, `go vet`, `go test`)
  - Added local guardrail commands via `Makefile` (`make ci`, `fmt-check`, `vet`, `test`)
  - Updated contributing docs
  - Added baseline metrics template (`planning/baseline/phase0-baseline.md`)

- **#12 Phase 1 – Execution safety hardening**
  - Added strict command parsing and allowlist policy for `run_command`
  - Removed shell execution path (`sh -c`), now executes direct binaries with args
  - Added workspace jail path checks for `read_file`/`write_file`
  - Normalized approval identity and approval replay protections
  - Added security tests for traversal, metacharacters/chaining, allow/deny, and approval normalization

- **#13 Phase 2 – Evaluation harness v1**
  - Added deterministic eval runner (`cmd/eval`) and schema (`eval/schema.go`)
  - Added 21 eval scenarios across tool-correctness, safety, memory fallback, and compaction flow
  - Added eval docs (`docs/eval-schema.md`)
  - Integrated eval job in CI and local `make eval`

- **#14 Phase 3 – Observability and run artifacts**
  - Added telemetry package with structured event schema
  - Added in-process metrics and artifact persistence (`metadata/transcript/tool summary/metrics`)
  - Added retention/pruning controls for artifacts
  - Integrated telemetry in agent/tool execution paths and memory/compaction decisions
  - Added telemetry tests
  - Added and wired env flags (`TELEMETRY_*`, `RUN_ARTIFACTS_*`)

- **#15 Phase 4 – Reliability and regression depth**
  - Added integration tests for multi-round tool flow and wallet-approval routing
  - Added failure-injection tests for embedding fallback and compaction fallback paths
  - Added queue panic-recovery regression test
  - Updated docs for deep reliability test subset

- **#16 Phase 5 – Benchmarking and release governance**
  - Added benchmark runner (`cmd/benchmark`) and config (`benchmarks/config.json`)
  - Added trend/report artifacts (JSON + markdown summary)
  - Added release gate command/script (`cmd/releasegate`, `scripts/release-gate.sh`)
  - Added CI benchmark/gate job and make targets (`make benchmark`, `make release-gate`)
  - Added release governance documentation and quality-impact changelog template

## Review/fix pass notes
A manual orchestrator review was run after each phase implementation. Notable fixes made during review included:
- keeping telemetry finalization consistent across all early-return paths,
- aligning telemetry metrics env flag naming and actual runtime behavior,
- removing generated test artifacts from tracked changes and ignoring benchmark/eval output directories.

## Validation
- `make ci`
- `make eval`
- `make benchmark`
- `make release-gate`

All checks pass on this branch.

<div><a href="https://cursor.com/agents/bc-9577c4ca-9cbe-47b3-aaf2-ec919b4eb142"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-9577c4ca-9cbe-47b3-aaf2-ec919b4eb142"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Review by fabriziogianni7 — COMMENTED** [review]
@cursoragent remove this planning folder as it’s not needed for the general project.

**Review by fabriziogianni7 — COMMENTED** [review]
@cursoragent remove this /planning folder as it’s not needed for the general project.

File: tools/tools.go [hunk 1]
```diff
@@ -1064,32 +1064,33 @@ func (t *Tools) httpRequest(args map[string]string, rawArgs map[string]interface
...
 		return "", fmt.Errorf("failed to load approvals: %w", err)
 	}
-	if IsApproved(command, approved) {
-		return executeCommand(command)
+	if IsApproved(parsed.Identity, approved) {
+		return executeCommand(parsed)
 	}
 
-	return "Permission denied. User can approve by saying 'approve: " + command + "' in chat.", nil
+	return "Permission denied. User can approve by saying 'approve: " + parsed.Identity + "' in chat.", nil
```
Comment (fabriziogianni7) [inline tools/tools.go:1089]: @cursoragent what is parsed.Identity here and where does it come from?
